src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -32,15 +38,91 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory, IDispo
         private readonly JoinableTaskCollection _deferredCleanupTasks;
 
+        /// <summary>
+        /// Cancellation token that controls existing async work that we have kicked off.  Canceled when we're finally
+        /// disposed.
+        /// </summary>
+        private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+
+        private readonly IForegroundNotificationService _notificationService;
+        private readonly IAsynchronousOperationListener _listener;
+        private readonly AsyncBatchingWorkQueue<DocumentId> _documentsToFireEventsFor;
+
         [ImportingConstructor]
         [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
-        public ProjectCodeModelFactory(VisualStudioWorkspace visualStudioWorkspace, [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, IThreadingContext threadingContext)
+        public ProjectCodeModelFactory(
+            VisualStudioWorkspace visualStudioWorkspace,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IThreadingContext threadingContext,
+            IForegroundNotificationService notificationService,
+            IAsynchronousOperationListenerProvider listenerProvider)
         {
             _visualStudioWorkspace = visualStudioWorkspace;
             _serviceProvider = serviceProvider;
             _threadingContext = threadingContext;
             _deferredCleanupTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
             _deferredCleanupTasks.DisplayName = nameof(ProjectCodeModelFactory) + "." + nameof(_deferredCleanupTasks);
+
+            _notificationService = notificationService;
+            _listener = listenerProvider.GetListener(FeatureAttribute.CodeModel);
+
+            // Queue up notifications we hear about docs changing.  that way we don't have to fire events multiple times
+            // for the same documents.  Once enough time has passed, take the documents that were changed and run
+            // through them, firing their latest events.
+            _documentsToFireEventsFor = new AsyncBatchingWorkQueue<DocumentId>(
+                TimeSpan.FromMilliseconds(visualStudioWorkspace.Options.GetOption(InternalSolutionCrawlerOptions.AllFilesWorkerBackOffTimeSpanInMS)),
+                ProcessNextDocumentBatchAsync,
+                // We only care about unique doc-ids, so pass in this comparer to collapse streams of changes for a
+                // single document down to one notification.
+                EqualityComparer<DocumentId>.Default,
+                _listener,
+                _cancellationTokenSource.Token);
+
+            _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
```

Comment:
💡 A using alias for `Task` would be helpful.

Reply:
Thought about it, but this is actually just totally clear to me what's going on.  If i needed to repeat this a ton i might feel differently :)